### PR TITLE
Wrap EvalScript in a ScriptExecution class

### PR DESCRIPTION
**Pull request by luke-jr**
When we last discussed making scripts debuggable (sometime after #3901), the plan was to instead trace execution rather than single-step through it.

This is the first step toward that goal. The full implementation can be found on my `script_debugger` branch.

### CODE REVIEW

File: src/script/interpreter.h [hunk 1]
```diff
@@ -174,7 +174,34 @@ class MutableTransactionSignatureChecker : public TransactionSignatureChecker
...
+    int nOpCount;
+
+    ScriptExecution(StackType& stack, const CScript&, unsigned int flags, const BaseSignatureChecker&, SigVersion);
+
+    bool Eval(ScriptError* error = NULL);
+};
+
+inline bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript& script, unsigned int flags, const BaseSignatureChecker& checker, SigVersion sigversion, ScriptError* error = NULL)
+{
+    return ScriptExecution(stack, script, flags, checker, sigversion).Eval(error);
```
Comment (promag) [inline src/script/interpreter.h]: Move `stack` to `::Eval()` for the time being as this is an output (also remove `stack` member from class)?

File: src/script/interpreter.h [hunk 2]
```diff
@@ -174,7 +174,34 @@ class MutableTransactionSignatureChecker : public TransactionSignatureChecker
     MutableTransactionSignatureChecker(const CMutableTransaction* txToIn, unsigned int nInIn, const CAmount& amountIn) : TransactionSignatureChecker(&txTo, nInIn, amountIn), txTo(*txToIn) {}
 };
 
-bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript& script, unsigned int flags, const BaseSignatureChecker& checker, SigVersion sigversion, ScriptError* error = NULL);
+class ScriptExecution {
```
Comment (promag) [inline src/script/interpreter.h]: Nit,rename to `ScriptExecuter`?

File: src/script/interpreter.h [hunk 3]
```diff
@@ -174,7 +174,34 @@ class MutableTransactionSignatureChecker : public TransactionSignatureChecker
...
+    CScript::const_iterator pc;
+    CScript::const_iterator pbegincodehash;
+
+    std::vector<bool> vfExec;
+    StackType altstack;
+    int nOpCount;
+
+    ScriptExecution(StackType& stack, const CScript&, unsigned int flags, const BaseSignatureChecker&, SigVersion);
+
+    bool Eval(ScriptError* error = NULL);
```
Comment (promag) [inline src/script/interpreter.h]: Rename to Execute?

File: src/script/interpreter.h (hunk 2 above)
Comment (luke-jr) [inline src/script/interpreter.h]: It can only execute a single script.

File: src/script/interpreter.h [hunk 4]
```diff
@@ -174,7 +174,34 @@ class MutableTransactionSignatureChecker : public TransactionSignatureChecker
     MutableTransactionSignatureChecker(const CMutableTransaction* txToIn, unsigned int nInIn, const CAmount& amountIn) : TransactionSignatureChecker(&txTo, nInIn, amountIn), txTo(*txToIn) {}
 };
 
-bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript& script, unsigned int flags, const BaseSignatureChecker& checker, SigVersion sigversion, ScriptError* error = NULL);
+class ScriptExecution {
+public:
+    typedef std::vector<unsigned char> StackElementType;
+    typedef std::vector<StackElementType> StackType;
+
+    const CScript& script;
```
Comment (luke-jr) [inline src/script/interpreter.h]: They are meant for external inspection.

File: src/script/interpreter.h (hunk 1 above)
Comment (luke-jr) [inline src/script/interpreter.h]: It's not only an output.

File: src/script/interpreter.h (hunk 4 above)
Comment (promag) [inline src/script/interpreter.h]: When that happens either make it public, create a getter or an `Inspect` method?

**Comment by luke-jr** [comment]
Rebased

**Comment by jtimon** [comment]
Concept ACK

File: src/script/interpreter.h [hunk 5]
```diff
@@ -185,7 +185,34 @@ class GenericTransactionSignatureChecker : public BaseSignatureChecker
...
+    SigVersion sigversion;
+
+    CScript::const_iterator pc;
+    CScript::const_iterator pbegincodehash;
+
+    std::vector<bool> vfExec;
+    StackType altstack;
+    int nOpCount;
+
+    ScriptExecution(StackType& stack, const CScript&, unsigned int flags, const BaseSignatureChecker&, SigVersion);
```
Comment (practicalswift) [inline src/script/interpreter.h]: Match parameter names between declaration and definition :-)

File: src/script/interpreter.h [hunk 6]
```diff
@@ -185,7 +185,34 @@ class GenericTransactionSignatureChecker : public BaseSignatureChecker
...
+
+    bool Eval(ScriptError* error = nullptr);
```
Comment (practicalswift) [inline src/script/interpreter.h]: Same here :-)

File: src/script/interpreter.cpp [hunk 7]
```diff
@@ -278,7 +278,12 @@ int FindAndDelete(CScript& script, const CScript& b)
     return nFound;
 }
 
-bool EvalScript(std::vector<std::vector<unsigned char> >& stack, const CScript& script, unsigned int flags, const BaseSignatureChecker& checker, SigVersion sigversion, ScriptError* serror)
+ScriptExecution::ScriptExecution(StackType& stack_in, const CScript& script_in, unsigned int flags_in, const BaseSignatureChecker& checker_in, SigVersion sigversion_in) :
+    script(script_in), stack(stack_in), flags(flags_in), checker(checker_in), sigversion(sigversion_in), pc(script.begin()), pbegincodehash(script.begin()), nOpCount(0)
```
Comment (practicalswift) [inline src/script/interpreter.cpp]: Should `vfExec` and `altstack` be initialized too?

**Comment by jtimon** [comment]
utACK, unless I'm missing something, this should not change behavior at all.

**Comment by luke-jr** [comment]
Abandoning this until there's some indication anyone cares

**Pull request closed without merging**